libraries/Microsoft.Bot.Builder/BotFrameworkStreamingExtensionsAdapter.cs

```diff
@@ -0,0 +1,695 @@
+
+            using (var context = new TurnContext(this, activity))
+            {
+                await RunPipelineAsync(context, callback, cancellationToken).ConfigureAwait(false);
+
+                if (activity.Type == ActivityTypes.Invoke)
+                {
+                    var activityInvokeResponse = context.TurnState.Get<Activity>(InvokeReponseKey);
+                    if (activityInvokeResponse == null)
+                    {
+                        return new InvokeResponse { Status = (int)HttpStatusCode.NotImplemented };
+                    }
+                    else
+                    {
+                        return (InvokeResponse)activityInvokeResponse.Value;
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Sends activities to the conversation.
+        /// Throws <see cref="ArgumentNullException"/> on null arguments.
+        /// Throws <see cref="ArgumentException"/> if activities length is zero.
+        /// </summary>
+        /// <param name="turnContext">The context object for the turn.</param>
+        /// <param name="activities">The activities to send.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        /// <remarks>If the activities are successfully sent, the task result contains
+        /// an array of <see cref="ResourceResponse"/> objects containing the IDs that
+        /// the receiving channel assigned to the activities.</remarks>
+        /// <seealso cref="ITurnContext.OnSendActivities(SendActivitiesHandler)"/>
+        public override async Task<ResourceResponse[]> SendActivitiesAsync(ITurnContext turnContext, Activity[] activities, CancellationToken cancellationToken)
+        {
+            if (turnContext == null)
+            {
+                throw new ArgumentNullException(nameof(turnContext));
+            }
+
+            if (cancellationToken == null)
+            {
+                throw new ArgumentNullException(nameof(cancellationToken));
+            }
+
+            if (activities == null)
+            {
+                throw new ArgumentNullException(nameof(activities));
+            }
+
+            if (activities.Length == 0)
+            {
+                throw new ArgumentException("Expecting one or more activities, but the array was empty.", nameof(activities));
+            }
+
+            var responses = new ResourceResponse[activities.Length];
+
+            /*
+             * NOTE: we're using for here (vs. foreach) because we want to simultaneously index into the
+             * activities array to get the activity to process as well as use that index to assign
+             * the response to the responses array and this is the most cost effective way to do that.
+             */
+            for (var index = 0; index < activities.Length; index++)
+            {
```

Comment:
Seeing this code again reminds me - this should be in a shared class, rather than copy/pasted. Please refactor accordingly.

Reply:
Again, I'd like to hold off on refactoring until we decide on a solid approach for the entire class hierarchy of the BotFramework adapters.